### ISSUE TITLE
make Federation + Dataloader work

### ISSUE THREAD

**Issue by Boshen** [async-graphql/async-graphql#431]
When Apollo Gateway sends something like 
```graphql
  query Query {
    _entities(representations: [{__typename:"User", userId:1},{__typename:"User", userId:2}]) {
      ...on User {
        userId
      }
    }
  }
```

dataloader inside `#[graphql(entity)]` is not working at the moment.

I guess the relevant code is here, but I couldn't investigate no further..:
```rust
                for item in representations {
                    res.push(
                        self.inner.find_entity(ctx, &item.0).await?.ok_or_else(|| {
                            ServerError::new("Entity not found.").at(ctx.item.pos)
                        })?,
                    );
                }
```

**Comment by sunli829** [comment]
I'll check this out tomorrow.😄

**Comment by sunli829** [comment]
I added an example:

https://github.com/async-graphql/async-graphql/blob/1e6ae23cd891ca91beb28a9cec71812bd7b8b24b/tests/federation.rs#L155

**Comment by Boshen** [comment]
@sunli829 I think you need to check the keys, otherwise your test passes despite dataloader malfunctioning:
```rust
        async fn load(&self, keys: &[ID]) -> Result<HashMap<ID, Self::Value>, Self::Error> {
            assert_eq!(
                vec![ID::from("1"), ID::from("2"), ID::from("3"), ID::from("4")],
                keys
            );
           ...
        }
```

I'm still learning async rust, but after some debugging it seems `Delay::new(self.delay).await;` inside `dataloader.load_many`is not blocking inside the `for item in representations` loop

**Comment by Boshen** [comment]
Using `join_all` worked:
```rust
                let representations: Vec<Any> = ctx.param_value("representations", None)?;
                let res = join_all(
                    representations
                        .iter()
                        .map(|item| self.inner.find_entity(ctx, &item.0)),
                )
                .await
                .into_iter()
                .map(|item| {
                    item.unwrap()
                        .ok_or_else(|| ServerError::new("Entity not found.").at(ctx.item.pos))
                        .unwrap()
                })
                .collect();
                return Ok(Some(Value::List(res)));
            }
```
But I guess there is a better way ...

**Comment by sunli829** [comment]
> Using `join_all` worked:
> 
> ```rust
>                 let representations: Vec<Any> = ctx.param_value("representations", None)?;
>                 let res = join_all(
>                     representations
>                         .iter()
>                         .map(|item| self.inner.find_entity(ctx, &item.0)),
>                 )
>                 .await
>                 .into_iter()
>                 .map(|item| {
>                     item.unwrap()
>                         .ok_or_else(|| ServerError::new("Entity not found.").at(ctx.item.pos))
>                         .unwrap()
>                 })
>                 .collect();
>                 return Ok(Some(Value::List(res)));
>             }
> ```
> 
> But I guess there is a better way ...

It is indeed better to use `join_all`, thank you for your suggestion. 🙂

**Comment by sunli829** [comment]
Released in `v2.5.11`

**Comment by Boshen** [comment]
🌹
